monitor: Enable to exec into the container

ktock opened the pull request:
Related to #1104

> PR6
> Add exec command to execute new processes in the same debug container. All processes are attachable as described in the previous section.

`exec` execs a new command in the container

```console
$ BUILDX_EXPERIMENTAL=1 /tmp/out/buildx build --invoke sh /tmp/ctx
...
(buildx) exec ps
Process "6zqrfs121s8fqo097mjf7j1t4" started. Press Ctrl-a-c to switch to that process.
(buildx) Switched IO


PID   USER     TIME  COMMAND
    1 root      0:00 sh
    7 root      0:00 ps
Switched IO
(buildx) exec ls
Process "57c0vhpkpo621nla15b2pj8oh" started. Press Ctrl-a-c to switch to that process.
(buildx) Switched IO


bin     etc     hello2  home    root    tmp     var
dev     hello   hello3  proc    sys     usr
Switched IO
(buildx) exec sh
Process "njnu3unnfb5zf5omot93yh6ap" started. Press Ctrl-a-c to switch to that process.
(buildx) Switched IO


/ # 
/ # ps
PID   USER     TIME  COMMAND
    1 root      0:00 sh
    8 root      0:00 sh
   14 root      0:00 ps
```

`attach` attaches IO to the specified process.

```console
$ BUILDX_EXPERIMENTAL=1 /tmp/out/buildx build --invoke sh /tmp/ctx
...
Interactive container was restarted with process "tmb96qnbc2q7xfjhochczysga". Press Ctrl-a-c to switch to the new container
(buildx) exec sh
Process "ooybngbcn0nnz584i8xvgbnii" started. Press Ctrl-a-c to switch to that process.
<C-a-c>
(buildx) exec sh
Process "xy51w8j7lzll5cduwfwdm7w50" started. Press Ctrl-a-c to switch to that process.
<C-a-c>
(buildx) exec sh
Process "yx8of20854f0jkd1ga3d9odan" started. Press Ctrl-a-c to switch to that process.
<C-a-c>
(buildx) ps
PID				CURRENT_SESSION	COMMAND
ooybngbcn0nnz584i8xvgbnii	false		[sh]
tmb96qnbc2q7xfjhochczysga	false		[sh]
xy51w8j7lzll5cduwfwdm7w50	false		[sh]
yx8of20854f0jkd1ga3d9odan	true		[sh]
(buildx) attach ooybngbcn0nnz584i8xvgbnii
Switched IO
Attached to process "ooybngbcn0nnz584i8xvgbnii". Press Ctrl-a-c to switch to the new container

/ # ps
PID   USER     TIME  COMMAND
    1 root      0:00 sh
    7 root      0:00 sh
   13 root      0:00 sh
   19 root      0:00 sh
   25 root      0:00 ps
/ # echo $$
7
/ # Switched IO
(buildx) attach tmb96qnbc2q7xfjhochczysga
Switched IO
Attached to process "tmb96qnbc2q7xfjhochczysga". Press Ctrl-a-c to switch to the new container

/ # echo $$
1
/ # Switched IO
(buildx) attach xy51w8j7lzll5cduwfwdm7w50
Switched IO
Attached to process "xy51w8j7lzll5cduwfwdm7w50". Press Ctrl-a-c to switch to the new container

/ # echo $$
13
/ # Switched IO
(buildx) attach yx8of20854f0jkd1ga3d9odan
Switched IO
Attached to process "yx8of20854f0jkd1ga3d9odan". Press Ctrl-a-c to switch to the new container

/ # echo $$
19
```

## note

- This commit includes a change to `docker-container` driver to always include init process in the driver container. This is for making sure to clean up exited processes created by BuildKit's container API.

jedevc reviewed (commented):
Shared some initial thoughts - but I think generally, it works! :tada: I'll definitely make sure that we don't stall on this PR as we did on #1296 :cry::cry:

Aside from the technical things, we definitely need to make sure we revisit the DX here (probably after all of https://github.com/docker/buildx/issues/1104 is completed) - the combination of commands we have some weird confusions, like `kill` kills the server instead of processes spawned with `exec` and seen with `ps`. I gave some suggestions in a comment on session lifetimes, so we may actually want to get rid of the explicit kill command in the server at some point (or maybe just rename it if we really want it).

ktock commented:
Thanks for the comments. Fixed the patch based on the comments.
Now processes are managed on the server side so monitor can switch IO among processes in different sessions without killing processes.

`exec` a process in a session:

```
(buildx) exec ps
Process "lg4atdbxa3al3l7l42lgo5sq3" started. Press Ctrl-a-c to switch to that process.
(buildx) Switched IO
PID   USER     TIME  COMMAND
    1 root      0:00 sh
    6 root      0:00 ps
Switched IO
(buildx) exec sh
Process "nwkw4tpg1zngwu1mkz600yrfe" started. Press Ctrl-a-c to switch to that process.
(buildx) Switched IO
/ # ps
PID   USER     TIME  COMMAND
    1 root      0:00 sh
   11 root      0:00 sh
   17 root      0:00 ps
/ # echo $$
11
```

`attach` to another process in the same session:

```
(buildx) ps
PID				CURRENT_SESSION	COMMAND
zncafsibzg3i6zriaviwsmzxw	false		[sh]
nwkw4tpg1zngwu1mkz600yrfe	true		[sh]
(buildx) Switched IO

/ # Switched IO
(buildx) attach zncafsibzg3i6zriaviwsmzxw
Attached to process "zncafsibzg3i6zriaviwsmzxw". Press Ctrl-a-c to switch to the new container
(buildx) Switched IO

/ # echo $$
1
```

`attach` to another process in another session:
The following attaches to process `3mqxm55ww9bcguidmkwmh96u2` in another session `1vo8nbia21t6rprjjtrmph3rv`.

```
(buildx) list
ID				CURRENT_SESSION
1vo8nbia21t6rprjjtrmph3rv	false
y0vvv1ta2bkpgt8osm0yhpzbd	true
(buildx) attach 1vo8nbia21t6rprjjtrmph3rv
Attached to process "". Press Ctrl-a-c to switch to the new container
(buildx) ps
PID				CURRENT_SESSION	COMMAND
3mqxm55ww9bcguidmkwmh96u2	false		[sh]
(buildx) attach 3mqxm55ww9bcguidmkwmh96u2
Attached to process "3mqxm55ww9bcguidmkwmh96u2". Press Ctrl-a-c to switch to the new container
(buildx) Switched IO

/ # echo $$
1
```

If other session already attached to this process, this will be automatically detached.

`attach` back to the original process in the original session (process `nwkw4tpg1zngwu1mkz600yrfe` in `y0vvv1ta2bkpgt8osm0yhpzbd`):

```
(buildx) list
ID				CURRENT_SESSION
1vo8nbia21t6rprjjtrmph3rv	true
y0vvv1ta2bkpgt8osm0yhpzbd	false
(buildx) attach y0vvv1ta2bkpgt8osm0yhpzbd
Attached to process "". Press Ctrl-a-c to switch to the new container
(buildx) ps
PID				CURRENT_SESSION	COMMAND
zncafsibzg3i6zriaviwsmzxw	false		[sh]
nwkw4tpg1zngwu1mkz600yrfe	false		[sh]
(buildx) attach nwkw4tpg1zngwu1mkz600yrfe
Attached to process "nwkw4tpg1zngwu1mkz600yrfe". Press Ctrl-a-c to switch to the new container
(buildx) Switched IO

/ # echo $$
11
```

ktock commented on controller/local/controller.go:
```diff
@@ -21,20 +21,41 @@ func NewLocalBuildxController(ctx context.Context, dockerCli command.Cli) contro
-	ccfg := build.ContainerConfig{
-		ResultCtx:  b.resultCtx,
+	if cfg.Rollback || b.container == nil || b.container.IsUnavailable() {
+		go b.curInvokeCancel()
```
Fixed.

ktock commented on controller/local/controller.go:
```diff
@@ -21,20 +21,41 @@ func NewLocalBuildxController(ctx context.Context, dockerCli command.Cli) contro
+	dockerCli        command.Cli
+	ref              string
+	resultCtx        *build.ResultContext
+	curInvokeCancels []func()
```
Fixed.

ktock commented on controller/remote/server.go:
```diff
@@ -27,16 +26,45 @@ func NewServer(buildFunc BuildFunc) *Server {
+	result             *build.ResultContext
+	inputPipe          *io.PipeWriter
+	curInvokeCancels   map[*build.InvokeConfig]func()
+	curInvokeCancelsMu sync.Mutex
```
Fixed.

ktock commented on build/build.go:
```diff
@@ -681,127 +679,214 @@ type ResultContext struct {
-		return errors.Errorf("result must be provided")
-	}
-	c, res := cfg.ResultCtx.Client, cfg.ResultCtx.Res
+type Container struct {
```
Fixed

ktock commented on build/build.go:
```diff
@@ -681,127 +679,214 @@ type ResultContext struct {
+			} else {
+				initRoutineStartedMu.Unlock()
+			}
+			wg.Wait()
```
Fixed.

ktock commented on build/build.go:
```diff
@@ -681,127 +679,214 @@ type ResultContext struct {
-			<-mainCtx.Done()
-			cancel()
-		}()
+	execFunc func(ctx context.Context, cfg *InvokeConfig) error
```
Fixed.

ktock commented on build/build.go:
```diff
@@ -681,127 +679,214 @@ type ResultContext struct {
-			}
-			close(doneCh)
-		}()
+func (c *Container) IsUnavailable() bool {
```
Overall SGTM. Fixed not to mark it as unavailable except 2 cases:

- 1: When PID1 exits. Currently PID1 is the first command executed by the user in the container (e.g. command specified by `--invoke`). In the future, maybe we can introduce a special PID1 process (e.g. tini?) for keeping the container available even when all user-specified commands exit.
- 2: When one of the processes returns non-zero exit code. In this case, BuildKit's container API kills the container. To fix this behaviour, we need to fix BuildKit first so we should do this as a long-term following up work.

ktock commented on monitor/monitor.go:
```diff
@@ -163,8 +173,66 @@ func RunMonitor(ctx context.Context, curRef string, options controllerapi.BuildO
+						curRef = ref
+					}
+					fmt.Fprintf(stdout, "Attached to process %q. Press Ctrl-a-c to switch to the new container\n", id)
+				case "exec":
```
Thank you for the suggestion. It can be a large refactoring so let's work in a following-up PR.

ktock commented on controller/pb/controller.proto:
```diff
@@ -123,10 +123,10 @@ message Message {
 
 message InitMessage {
   string Ref = 1;
-  ContainerConfig ContainerConfig = 2;
```
Fixed to manage process on the server side so monitor can switch IO among processes in different sessions without killing processes.

ktock commented on build/build.go:
```diff
@@ -681,127 +679,214 @@ type ResultContext struct {
-			return nil, err
-		}
-		defer ctr.Release(context.TODO())
+func NewContainer(ctx context.Context, resultCtx *ResultContext) (*Container, error) {
```
Fixed.

jedevc commented on monitor/monitor.go:
```diff
@@ -198,45 +270,74 @@ type readWriter struct {
+		if err := m.invoke(ctx, ref, pid, cfg); err != nil {
 			logrus.Debugf("invoke error: %v", err)
 		}
+		m.attachedPidMu.Lock()
+		if pid == m.attachedPid {
+			m.attachedPid = ""
+		}
+		m.attachedPidMu.Unlock()
```
We can use https://pkg.go.dev/sync/atomic@master#Value here. There's a few other places in this PR that we should use the atomic primitives for where we would otherwise need a mutex to access a single value.

ktock replied:
Fixed.

jedevc reviewed (commented):
Thanks @ktock :heart: This already looks a lot cleaner to me, just had a few more minor comments.

I noticed you also opened https://github.com/docker/buildx/pull/1640, which seems to complete https://github.com/docker/buildx/issues/1104? I'll take a look at that one asap as well, but ideally before moving forward past that into the next steps, we can dive into the `TODO`s left in the code as well as tidy up the lifetimes stuff, and clean up the UX before going too much further? I don't want to block merging this or any of the other debug PRs in, but I do want to make sure we have a strong foundation before trying to add too much more functionality (like breakpoints, etc).

ktock commented on build/invoke.go:
```diff
@@ -0,0 +1,248 @@
+func (c *Container) IsUnavailable() bool {
+	c.isUnavailableMu.Lock()
+	defer c.isUnavailableMu.Unlock()
+	return c.isUnavailable
```
Fixed

ktock commented on build/invoke.go:
```diff
@@ -0,0 +1,248 @@
+}
+
+func (c *Container) Exec(ctx context.Context, cfg *InvokeConfig) error {
+	killCh := make(chan struct{})
```
Fixed.

ktock commented on build/invoke.go:
```diff
@@ -0,0 +1,248 @@
+	c.initStartedChMu.Lock()
+	initCh := c.initStartedCh
+	c.initStartedCh = nil
+	c.initStartedChMu.Unlock()
```
Fixed.

ktock commented on controller/control/controller.go:
```diff
@@ -9,12 +9,18 @@ import (
+	// Invoke starts an IO session into the specified process.
+	// If pid doesn't matche to any running processes, it starts a new process with the specified config.
+	// If there is no container running or InvokeConfig.Rollback is speicfied, the process will start in a newly created container.
+	// NOTE: If needed, in the future, we can split this API into three APIs (NewContainer, NewProcess and Attach).
```
Thanks for the suggestion. Let's work on the refactor on this API in the following-up PRs.

ktock commented on controller/processes/processes.go:
```diff
@@ -0,0 +1,168 @@
+			m.processes.Delete(pid)
+			processCancelFunc()
+			p.errCh <- err
+		}()
```
Fixed

ktock commented on controller/remote/server.go:
```diff
@@ -27,16 +27,45 @@ func NewServer(buildFunc BuildFunc) *Server {
+	statusChan  chan *client.SolveStatus
+	result      *build.ResultContext
+	inputPipe   *io.PipeWriter
+	cancelBuild func()
```
No it isn't. Fixed to return an error if a build happens during another ongoing.

ktock commented on monitor/monitor.go:
```diff
@@ -70,18 +74,17 @@ func RunMonitor(ctx context.Context, curRef string, options controllerapi.BuildO
 			if prev == 0 && res == 0 {
 				// No toggle happened because container I/O isn't enabled.
-				return "No running interactive containers. You can start one by issuing rollback command\n"
+				return "Process isn't attachd. Use \"attach\" for attaching or \"rollback\" or \"exec\" for running new one.\n"
```
Fixed

ktock commented on monitor/monitor.go:
```diff
@@ -163,8 +179,68 @@ func RunMonitor(ctx context.Context, curRef string, options controllerapi.BuildO
+					pid := m.exec(ctx, curRef, cfg)
+					fmt.Fprintf(stdout, "Process %q started. Press Ctrl-a-c to switch to that process.\n", pid)
+				case "ps":
+					plist, err := c.ListProcesses(ctx, curRef)
+					if err != nil {
+						fmt.Println("cannot list process:", err)
+						continue
+					}
```
Thanks for the suggestion. SGTM.

ktock commented:
@jedevc Thank you for the review. Fixed the patch based on the comments,

ktock commented:
> I noticed you also opened https://github.com/docker/buildx/pull/1640, which seems to complete https://github.com/docker/buildx/issues/1104? I'll take a look at that one asap as well, but ideally before moving forward past that into the next steps, we can dive into the TODOs left in the code as well as tidy up the lifetimes stuff, and clean up the UX before going too much further? I don't want to block merging this or any of the other debug PRs in, but I do want to make sure we have a strong foundation before trying to add too much more functionality (like breakpoints, etc).

SGTM :+1:

jedevc commented on build/invoke.go:
```diff
@@ -0,0 +1,229 @@
+)
+
+// InvokeConfig is configuration for a process to run in the container.
+type InvokeConfig struct {
```
I think we should remove this struct, and instead use `controllerapi.InvokeConfig`, it would be good to avoid having multiple duplicate struct. The main difference seems to be that we obviously can't attach `io` objects to it, we can pass those as separate args (maybe using `ioset.In` to stick them together).

Note on having the duplicate `User` and `NoUser` fields to represent an optional value (which if we were using a modern version of protobuf, we could use the `optional` field in https://github.com/protocolbuffers/protobuf/blob/main/docs/field_presence.md). I guess there's not reason we have to use gogo proto for buildx like we do for buildkit (we don't care about version compatibility), we could just switch to using upstream? https://pkg.go.dev/google.golang.org/protobuf. Definitely follow-up material :tada: cc @crazy-max since you've done protobuf investigations.

ktock replied:
Thank you for the comment. Removed `InvokeConfig`.

jedevc commented on controller/local/controller.go:
```diff
@@ -36,38 +40,63 @@ func (b *localController) Build(ctx context.Context, options controllerapi.Build
 }
 
 func (b *localController) Kill(context.Context) error {
 	return nil // nop
```
We probably want a call to `b.cancelRunningProcesses()` here as well?

ktock replied:
Thank you for the comment. Added b.cancelRunningProcesses()

jedevc commented on controller/local/controller.go:
```diff
@@ -36,38 +40,63 @@ func (b *localController) Build(ctx context.Context, options controllerapi.Build
+func (b *localController) ListProcesses(ctx context.Context, ref string) (infos []*controllerapi.ProcessInfo, retErr error) {
 	if ref != b.ref {
-		return errors.Errorf("unknown ref %q", ref)
+		return nil, errors.Errorf("unknown ref %q", ref)
```
I assume that the local controller also doesn't support multiple build requests at a time? We should add an explicit error in this case as well, similar to https://github.com/docker/buildx/pull/1626#discussion_r1112960386.

ktock replied:
Fixed to return error on parallel builds.

jedevc commented:
These should just be nits now, any major changes we can do in follow-ups :+1:

jedevc reviewed (approved):
LGTM, happy to address the unresolved comments as follow-ups later.

PTAL @tonistiigi :smile:

ktock commented:
@tonistiigi PTAL :pray:
